### PR TITLE
Dev 6527.b

### PR DESCRIPTION
**Pull request by steco**
# Pull Request Checklist

- [x] Read the [contributing guidelines](https://github.com/finbourne/lusid-sdk-python-preview/blob/master/CONTRIBUTING.md)
- [x] Tests pass
- [x] Raised the PR against the `develop` branch

# Description of the PR

Describe the code changes for the reviewers, explain the solution you have provided and how it fixes the issue

### CODE REVIEW

File: sdk/lusid/utilities/api_client_builder.py [hunk 1]
```diff
@@ -19,26 +21,50 @@ def build(api_secrets_filename=None, okta_response_handler=None, api_configurati
...
+        # extra request args
+        kwargs = {"headers": headers}
+
+        # use proxy if supplied
+        proxy_config = configuration.proxy_config
+        if proxy_config is not None:
+
+            proxy_url = proxy_config.address
+            if proxy_config.username is not None and proxy_config.password is not None:
+                index = proxy_config.address.index("://")
```
Comment (steco) [inline sdk/lusid/utilities/api_client_builder.py]: Are you sure we should be massaging the values like this?

File: sdk/lusid/utilities/api_client_builder.py [hunk 2]
```diff
@@ -56,4 +82,13 @@ def build(api_secrets_filename=None, okta_response_handler=None, api_configurati
         config.access_token = api_token
         config.host = configuration.api_url
 
+        # set the cert
+        config.ssl_ca_cert = configuration.certificate_filename
+
+        # set the poxy for LUSID if needed
```
Comment (steco) [inline sdk/lusid/utilities/api_client_builder.py]: typo

File: sdk/lusid/utilities/refreshing_token.py [hunk 3]
```diff
@@ -39,7 +40,12 @@ def get_refresh_token():
                 }
 
                 request_body = f"grant_type=refresh_token&scope=openid client groups offline_access&refresh_token={refresh_token}"
-                okta_response = requests.post(token_url, data=request_body, headers=headers)
+
+                if certificate_filename is not None:
```
Comment (steco) [inline sdk/lusid/utilities/refreshing_token.py]: Is this `if` needed?

File: sdk/lusid/utilities/api_client_factory.py [hunk 4]
```diff
@@ -20,23 +20,35 @@ def __init__(self, **kwargs):
...
+        if "certificate_filename" in kwargs.keys():
+            certificate_filename = str(kwargs["certificate_filename"])
+
         if ("token" in kwargs and str(kwargs["token"]) != "None") and api_url is not None:
 
             config = lusid.Configuration()
             config.access_token = kwargs["token"]
             config.host = api_url
 
+            if certificate_filename is not None:
```
Comment (steco) [inline sdk/lusid/utilities/api_client_factory.py]: This is duplicated in ApiClientBuilder, where it is slightly different.  Can the code be consolidated?

File: sdk/lusid/utilities/api_client_builder.py (hunk 2 above)
Comment (rizwansaeed) [reply sdk/lusid/utilities/api_client_builder.py]: fixed

File: sdk/lusid/utilities/refreshing_token.py (hunk 3 above)
Comment (rizwansaeed) [reply sdk/lusid/utilities/refreshing_token.py]: has been refactored in latest version

File: sdk/lusid/utilities/api_client_builder.py (hunk 1 above)
Comment (rizwansaeed) [reply sdk/lusid/utilities/api_client_builder.py]: this is substituting the username/password into a proxy url format (if they are supplied), open to suggestions

(note this has been refactored in the latest revision)

File: sdk/docker-compose.yml [hunk 5]
```diff
@@ -2,6 +2,18 @@ version: "3"
 
 services:
 
+  tinyproxy:
+    image: dannydirect/tinyproxy:latest
```
Comment (steco) [inline sdk/docker-compose.yml]: Should we use `dannydirect`?

Comment (rizwansaeed) [reply sdk/docker-compose.yml]: Its only for the tests in the travis build, I think its ok

File: sdk/lusid/utilities/api_client_factory.py [hunk 6]
```diff
@@ -17,39 +17,43 @@ def __init__(self, **kwargs):
...
+        if "token" in kwargs and str(kwargs["token"]) != "None":
+            # If there is a token use it along with the specified proxy details if specified
+            config = ApiConfiguration(
+                api_url=kwargs.get("api_url", os.getenv("FBN_LUSID_API_URL", None)),
+                token_url=None,
+                username=None,
+                password=None,
+                client_id=None,
+                client_secret=None,
+                certificate_filename=kwargs.get("certificate_filename", None),
```
Comment (rizwansaeed) [inline sdk/lusid/utilities/api_client_factory.py]: should this also try the env var `FBN_CLIENT_CERTIFICATE`?

File: sdk/lusid/utilities/refreshing_token.py [hunk 7]
```diff
@@ -39,7 +41,12 @@ def get_refresh_token():
                 }
 
                 request_body = f"grant_type=refresh_token&scope=openid client groups offline_access&refresh_token={refresh_token}"
-                okta_response = requests.post(token_url, data=request_body, headers=headers)
+
+                # request parameters
+                kwargs = {"headers": headers}
+                kwargs["proxies"] = proxies
```
Comment (rizwansaeed) [inline sdk/lusid/utilities/refreshing_token.py]: Is it ok setting `kwargs["proxies"]` to `None`?

**Pull request merged**